tests/slow/src/test/scala/tests/mill/MillServerSuite.scala

```diff
@@ -24,11 +24,22 @@ class MillServerSuite
   val supportedBspVersion = V.millVersion
   val scalaVersion = V.scala213
   val buildTool: MillBuildTool = MillBuildTool(() => userConfig)
+  def bspTrace: AbsolutePath =
+    workspace.resolve(".metals").resolve("bsp.trace.json")
 
   override def currentDigest(
       workspace: AbsolutePath
   ): Option[String] = MillDigest.current(workspace)
 
+  override def afterEach(context: AfterEach): Unit = {
+    if (context.test.name == s"generate-and-connect-$supportedBspVersion") {
```

Comment:
The trace needs to be create first, so you would also need to implement afterEach. Maybe instead you could just create a method `testWithBsp` that would wrap the test itself and create the bsp trace before and print it in a catch clause?

This way we could reuse the method and not need to match on the test name

Reply:
The problem is that I want to print it even on timeout, which is put on the whole test. I think only doing it `after` the test will work in that situation. I guess I can not use using `BaseFunSuite.test` and instead add the test to buffer by hand...?

Reply:
We have a very similar thing in `BaseDapSuite`, you can add a test transform by overriding munitTestTransforms